### PR TITLE
Users/shreyas r msft/input parser migration

### PR DESCRIPTION
**Pull request by ShreyasRmsft**

### CODE REVIEW

File: Tasks/VsTest/models.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import * as version from './vstestversion';
+import { InputDataContract } from './inputdatacontract';
```
Comment (ShreyasRmsft) [inline Tasks/VsTest/models.ts]: will remove this

File: Tasks/VsTest/taskinputparser.ts [hunk 2]
```diff
@@ -9,56 +9,422 @@ import * as os from 'os';
...
-    if (dtaConfiguration.vsTestLocationMethod === utils.Constants.vsTestVersionString && dtaConfiguration.vsTestVersion === '12.0') {
-        throw (tl.loc('vs2013NotSupportedInDta'));
+    inputDataContract.DistributionSettings.NumberOfTestAgents = 1;
+    const totalJobsInPhase = parseInt(tl.getVariable('SYSTEM_TOTALJOBSINPHASE'));
+    if (!isNaN(totalJobsInPhase)) {
+        inputDataContract.DistributionSettings.NumberOfTestAgents = totalJobsInPhase;
     }
+    console.log(tl.loc('dtaNumberOfAgents', inputDataContract.DistributionSettings.NumberOfTestAgents));
+
+    const distributionType = tl.getInput('distributionBatchType');
```
Comment (ShreyasRmsft) [inline Tasks/VsTest/taskinputparser.ts]: move this to dta

File: Tasks/VsTest/taskinputparser.ts [hunk 3]
```diff
@@ -9,56 +9,422 @@ import * as os from 'os';
...
+
+    inputDataContract.DistributionSettings = <inputdatacontract.DistributionSettings>{};
+
+    if (distributionType && distributionType === 'basedOnTestCases') {
+        inputDataContract.DistributionSettings.TestCaseLevelSlicingEnabled = true;
+        // flow if the batch type = based on agents/custom batching
+        const distributeByAgentsOption = tl.getInput('batchingBasedOnAgentsOption');
+        if (distributeByAgentsOption && distributeByAgentsOption === 'customBatchSize') {
+            const batchSize = parseInt(tl.getInput('customBatchSizeValue'));
+            if (!isNaN(batchSize) && batchSize > 0) {
```
Comment (ShreyasRmsft) [inline Tasks/VsTest/taskinputparser.ts]: do we want to do thjis here?

File: Tasks/VsTest/taskinputparser.ts [hunk 4]
```diff
@@ -9,56 +9,422 @@ import * as os from 'os';
...
-    return dtaConfiguration;
+
+    // Rerun information
+    inputDataContract.ExecutionSettings.RerunSettings.RerunFailedTests = tl.getBoolInput('rerunFailedTests');
+    console.log(tl.loc('rerunFailedTests', inputDataContract.ExecutionSettings.RerunSettings.RerunFailedTests));
+
+    const rerunType = tl.getInput('rerunType') || 'basedOnTestFailurePercentage';
+
+    // hydra: unravel the nestings
+    if (rerunType === 'basedOnTestFailureCount') {
```
Comment (ShreyasRmsft) [inline Tasks/VsTest/taskinputparser.ts]: switching off rerun logic to dta executiion host?

File: Tasks/VsTest/distributedtest.ts [hunk 5]
```diff
@@ -31,8 +31,10 @@ export class DistributedTest {
     }
 
     private publishCodeChangesIfRequired(): void {
-        if (this.dtaTestConfig.tiaConfig.tiaEnabled) {
-            const code = testSelector.publishCodeChanges(this.dtaTestConfig.tiaConfig, this.dtaTestConfig.proxyConfiguration, null, this.dtaTestConfig.taskInstanceIdentifier);
+        if (this.inputDataContract.ExecutionSettings.TiaSettings.Enabled) {
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: null check

File: Tasks/VsTest/distributedtest.ts [hunk 6]
```diff
@@ -54,208 +56,81 @@ export class DistributedTest {
...
-        inputDataContract.TfsSpecificSettings = <inputdatacontract.TfsSpecificSettings>{};
-        inputDataContract.TfsSpecificSettings.BuildId = utils.Helper.isNullEmptyOrUndefined(tl.getVariable('Build.Buildid')) ? null : Number(tl.getVariable('Build.Buildid'));
-        inputDataContract.TfsSpecificSettings.BuildUri = tl.getVariable('Build.BuildUri');
-        inputDataContract.TfsSpecificSettings.ReleaseId = utils.Helper.isNullEmptyOrUndefined(tl.getVariable('Release.ReleaseId')) ? null : Number(tl.getVariable('Release.ReleaseId'));
-        inputDataContract.TfsSpecificSettings.ReleaseUri = tl.getVariable('Release.ReleaseUri');
+        // let envVars: { [key: string]: string; } = <{ [key: string]: string; }>{};
+        const envVars: { [key: string]: string; } = process.env; // This is a temporary solution where we are passing parent process env vars, we should get away from this
+        this.inputDataContract.TestSelectionSettings.TestSourcesFile = this.createTestSourcesFile();
+        // Temporary solution till this logic can move to the test platform itself
+        if (this.inputDataContract.UsingXCopyTestPlatformPackage) {
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: seperate method

File: Tasks/VsTest/distributedtest.ts [hunk 7]
```diff
@@ -54,208 +56,81 @@ export class DistributedTest {
...
-                utils.Helper.printMultiLineLog(settings, (logLine) => { console.log('##vso[task.debug]' + logLine); });
-            });
         }
 
         // Pass the acess token as an environment variable for security purposes
-        // let envVars: { [key: string]: string; } = <{ [key: string]: string; }>{};
-        const envVars: { [key: string]: string; } = process.env; // This is a temporary solutio where we are passing parent process env vars, we should get away from this
-        utils.Helper.addToProcessEnvVars(envVars, 'DTA.AccessToken', this.dtaTestConfig.dtaEnvironment.patToken);
+        utils.Helper.addToProcessEnvVars(envVars, 'DTA.AccessToken', this.inputDataContract.AccessToken);
+        this.inputDataContract.AccessToken = null;
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: not required?

File: Tasks/VsTest/distributedtest.ts [hunk 8]
```diff
@@ -54,208 +56,81 @@ export class DistributedTest {
...
 
         // Invoke DtaExecutionHost with the input json file
         const inputFilePath = path.join(tl.getVariable('temp'), 'input.json');
-        DistributedTest.removeEmptyNodes(inputDataContract);
-        writeFileSync(inputFilePath, JSON.stringify(inputDataContract));
+        DistributedTest.removeEmptyNodes(this.inputDataContract);
+        writeFileSync(inputFilePath, JSON.stringify(this.inputDataContract));
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: meaningful error? and telemetry

File: Tasks/VsTest/distributedtest.ts [hunk 9]
```diff
@@ -290,8 +165,8 @@ export class DistributedTest {
 
     private cleanUpDtaExeHost() {
         try {
-            if (this.testSourcesFile) {
-                tl.rmRF(this.testSourcesFile);
+            if (this.inputDataContract.TestSelectionSettings.TestSourcesFile) {
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: do we really require as now we are using AGENT.TEMP?

File: Tasks/VsTest/testselectorinvoker.ts [hunk 10]
```diff
@@ -106,6 +107,109 @@ export class TestSelectorInvoker {
         return output.code;
     }
 
+    public publishCodeChangesInDistributedMode(inputDataContract: inputdatacontract.InputDataContract): number {
```
Comment (nigurr) [inline Tasks/VsTest/testselectorinvoker.ts]: get it reviewed with praga

File: Tasks/VsTest/distributedtest.ts (hunk 7 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/distributedtest.ts]: setting it back to null. This is required

File: Tasks/VsTest/distributedtest.ts (hunk 9 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/distributedtest.ts]: removed this

File: Tasks/VsTest/distributedtest.ts [hunk 11]
```diff
@@ -301,7 +176,13 @@ export class DistributedTest {
 
     private createTestSourcesFile(): string {
         try {
-            const sources = tl.findMatch(this.dtaTestConfig.testDropLocation, this.dtaTestConfig.sourceFilter);
+            let sourceFilter = tl.getDelimitedInput('testAssemblyVer2', '\n', true);
+
+            if (this.inputDataContract.TestSelectionSettings.TestSelectionType.toLowerCase() !== 'testassemblies') {
+                sourceFilter = ['**\\*', '!**\\obj\\*'];
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply Tasks/VsTest/distributedtest.ts]: We had been doing this replication from some time now. Will have to go ahead and remove it from the C# layer

File: Tasks/VsTest/task.json [hunk 12]
```diff
@@ -16,8 +16,8 @@
     "author": "Microsoft Corporation",
     "version": {
         "Major": 2,
-        "Minor": 5,
-        "Patch": 6
+        "Minor": 134,
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply Tasks/VsTest/task.json]: was considering setting it to sprint number. Just a thought

File: Tasks/VsTest/inputparser.ts [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+
+    inputDataContract.CollectionUri = tl.getVariable('System.TeamFoundationCollectionUri');
+    inputDataContract.AccessToken = tl.getEndpointAuthorization('SystemVssConnection', true).parameters['AccessToken'];
+    inputDataContract.AgentName = tl.getVariable('Agent.MachineName') + '-' + tl.getVariable('Agent.Name') + '-' + tl.getVariable('Agent.Id');
+
+    //hydra: change this input to be some unique run identifier
+    inputDataContract.RunIdentifier = getRunIdentifier();
+
+    //hydra: do we need
+    //dtaEnvironment.dtaHostLogFilePath = path.join(tl.getVariable('System.DefaultWorkingDirectory'), 'DTAExecutionHost.exe.log');
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: remove this

File: Tasks/VsTest/inputparser.ts [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+
+    return inputDataContract;
+}
+
+function getTestSelectionInputs(inputDataContract : idc.InputDataContract) : idc.InputDataContract {
+    inputDataContract.TestSelectionSettings = <idc.TestSelectionSettings>{};
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: check this works in Node 5?

File: Tasks/VsTest/inputparser.ts [hunk 15]
```diff
@@ -0,0 +1,481 @@
...
+
+    return inputDataContract;
+}
+
+function getTestPlatformSettings(inputDataContract : idc.InputDataContract) : idc.InputDataContract {
+    const vsTestLocationMethod = tl.getInput('vstestLocationMethod');
+    if (vsTestLocationMethod === utils.Constants.vsTestVersionString) {
+        const vsTestVersion = tl.getInput('vsTestVersion');
+        if (utils.Helper.isNullEmptyOrUndefined(vsTestVersion)) {
+            console.log('vsTestVersion is null or empty');
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: localize

File: Tasks/VsTest/inputparser.ts [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+        } else if (vsTestLocationMethod === utils.Constants.vsTestVersionString && vsTestVersion === '12.0') {
+            throw (tl.loc('vs2013NotSupportedInDta'));
+        } else {
+            console.log(tl.loc('vsVersionSelected', vsTestVersion));
+            inputDataContract.VsTestConsolePath = getTestPlatformPath(inputDataContract);
+        }
+    } else {
+        // hydra: should it be full path or directory above?
+        inputDataContract.VsTestConsolePath = tl.getInput('vsTestLocation');
+        console.log(tl.loc('vstestLocationSpecified', 'vstest.console.exe', inputDataContract.VsTestConsolePath));
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: 1. VsTools path -> check what's the path
2. Vstest.console.exe path -> vs 2015 [IDE]
3. vstest.console.exe path -> vs 15.5
4. vstest.console.exe path -> 15.6 > [\Extensions\TestWindow\]

File: Tasks/VsTest/inputparser.ts [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+    const totalJobsInPhase = parseInt(tl.getVariable('SYSTEM_TOTALJOBSINPHASE'));
+    if (!isNaN(totalJobsInPhase)) {
+        inputDataContract.DistributionSettings.NumberOfTestAgents = totalJobsInPhase;
+    }
+    console.log(tl.loc('dtaNumberOfAgents', inputDataContract.DistributionSettings.NumberOfTestAgents));
+
+    const distributionType = tl.getInput('distributionBatchType');
+
+    if (distributionType && distributionType === 'basedOnTestCases') {
+        inputDataContract.DistributionSettings.TestCaseLevelSlicingEnabled = true;
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: do the naming change

File: Tasks/VsTest/inputparser.ts [hunk 18]
```diff
@@ -0,0 +1,481 @@
...
+    }
+    console.log(tl.loc('runSettingsFileInput', inputDataContract.ExecutionSettings.SettingsFile));
+
+    inputDataContract.ExecutionSettings.OverridenParameters = tl.getInput('overrideTestrunParameters');
+
+    inputDataContract.ExecutionSettings.AssemblyLevelParallelism = tl.getBoolInput('runInParallel');
+    console.log(tl.loc('runInParallelInput', inputDataContract.ExecutionSettings.AssemblyLevelParallelism));
+
+    // hydra: valid only for console flow
+    //inputDataContract.ExecutionSettings.runTestsInIsolation = tl.getBoolInput('runTestsInIsolation');
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: these should exist

File: Tasks/VsTest/inputparser.ts [hunk 19]
```diff
@@ -0,0 +1,481 @@
...
+    tl.debug('ProceedAfterAbortedTestCase is set to : ' + inputDataContract.ExecutionSettings.ProceedAfterAbortedTestCase);
+
+    // hydra: Maybe move all warnings to a diff function
+    if (tl.getBoolInput('uiTests') && inputDataContract.ExecutionSettings.AssemblyLevelParallelism) {
+        tl.warning(tl.loc('uitestsparallel'));
+    }
+
+    // hydra: console flow only
+    //testConfiguration.otherConsoleOptions = tl.getInput('otherConsoleOptions');
+    //console.log(tl.loc('otherConsoleOptionsInput', testConfiguration.otherConsoleOptions));\
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: keep them

File: Tasks/VsTest/inputparser.ts [hunk 20]
```diff
@@ -0,0 +1,481 @@
...
+        return inputDataContract;
+    }
+
+    inputDataContract.ExecutionSettings.TiaSettings = <idc.TiaSettings>{};
+    inputDataContract.ExecutionSettings.TiaSettings.Enabled = tl.getBoolInput('runOnlyImpactedTests');
+    inputDataContract.ExecutionSettings.TiaSettings.RebaseLimit = +tl.getInput('runAllTestsAfterXBuilds');
+    inputDataContract.ExecutionSettings.TiaSettings.FileLevel = getTIALevel(tl.getVariable('tia.filelevel'));
+    inputDataContract.ExecutionSettings.TiaSettings.SourcesDirectory = tl.getVariable('build.sourcesdirectory');
+    inputDataContract.ExecutionSettings.TiaSettings.FilterPaths = tl.getVariable('TIA_IncludePathFilters');
+    // hydra: inputDataContract.ExecutionSettings.TiaSettings.runIdFile = path.join(os.tmpdir(), uuid.v1() + '.txt');
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: unless testselector tool code is fixed.

File: Tasks/VsTest/inputparser.ts [hunk 21]
```diff
@@ -0,0 +1,481 @@
...
+}
+
+function getTestPlatformPath(inputDataContract : idc.InputDataContract) {
+    let vsTestVersion = tl.getInput('vsTestVersion');
+    if (vsTestVersion.toLowerCase() === 'latest') {
+        // latest
+        tl.debug('Searching for latest Visual Studio');
+        const vstestconsole15Path = versionfinder.getVSTestConsole15Path();
+        if (vstestconsole15Path) {
+            vsTestVersion = '15.0';
```
Comment (nigurr) [inline Tasks/VsTest/inputparser.ts]: remove line

Comment (nigurr) [reply Tasks/VsTest/inputparser.ts]: check for manual vstest.console.exe path where cx can give full path or dir

File: Tasks/VsTest/task.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (prawalagarwal) [reply Tasks/VsTest/task.json]: I vote for sprint number too

File: Tasks/VsTest/testselectorinvoker.ts (hunk 10 above)
Comment (prawalagarwal) [inline Tasks/VsTest/testselectorinvoker.ts]: Help me understand this. Why did we create a new invoker for distributed?

File: Tasks/VsTest/distributedtest.ts (hunk 5 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/distributedtest.ts]: done

File: Tasks/VsTest/inputparser.ts [hunk 22]
```diff
@@ -0,0 +1,481 @@
...
+    inputDataContract = getTestReportingSettings(inputDataContract);
+    inputDataContract = getTestPlatformSettings(inputDataContract);
+    inputDataContract = getLoggingSettings(inputDataContract);
+    inputDataContract = getProxySettings(inputDataContract);
+    inputDataContract = getDistributionSettings(inputDataContract);
+    inputDataContract = getExecutionSettings(inputDataContract);
+
+    taskinputparser.logWarningForWER(tl.getBoolInput('uiTests'));
+
+    inputDataContract.UseNewCollector = false;
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: temp var. leaving it as it is for now. Will remove in the next testselector pr

File: Tasks/VsTest/inputparser.ts [hunk 23]
```diff
@@ -0,0 +1,481 @@
...
+        inputDataContract.IsPrFlow = true;
+    } else {
+        inputDataContract.IsPrFlow = utils.Helper.stringToBool(tl.getVariable('tia.isPrFlow'));
+    }
+    inputDataContract.UseTestCaseFilterInResponseFile = utils.Helper.stringToBool(tl.getVariable('tia.useTestCaseFilterInResponseFile'));
+
+    inputDataContract.TeamProject = tl.getVariable('System.TeamProject');
+    inputDataContract.VstestTaskInstanceIdentifier = uuid.v1();
+
+    inputDataContract.UseVsTestConsole = false;
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 13 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 14 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 15 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 17 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 18 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: added back

File: Tasks/VsTest/inputparser.ts (hunk 19 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done

File: Tasks/VsTest/inputparser.ts (hunk 20 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: not being used today.

File: Tasks/VsTest/inputparser.ts (hunk 21 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/inputparser.ts]: done in the invoking function above

File: Tasks/VsTest/task.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply Tasks/VsTest/task.json]: will ask shiva and finalize

File: Tasks/VsTest/testselectorinvoker.ts (hunk 10 above)
Comment (ShreyasRmsft) [reply Tasks/VsTest/testselectorinvoker.ts]: because inputs need to come from inputdatacontract which currently does not exist for single agent flow.

**Pull request merged**